pallets/circuit-breaker/src/lib.rs

```diff
@@ -0,0 +1,469 @@
+// This file is part of HydraDX.
+
+// Copyright (C) 2020-2022  Intergalactic, Limited (GIB).
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::{Decode, Encode};
+use frame_support::{ensure, pallet_prelude::DispatchResult, traits::Get};
+use hydradx_traits::{OnLiquidityChangeHandler, OnPoolStateChangeHandler};
+use scale_info::TypeInfo;
+use sp_core::MaxEncodedLen;
+use sp_runtime::traits::{AtLeast32BitUnsigned, CheckedAdd, CheckedDiv, CheckedMul, CheckedSub, Zero};
+use sp_runtime::{ArithmeticError, DispatchError, RuntimeDebug};
+
+pub mod weights;
+
+#[cfg(any(feature = "runtime-benchmarks", test))]
+mod benchmarking;
+
+#[cfg(test)]
+mod tests;
+
+/// Max trade volume limit multiplier of liquidity that can be traded in a block
+pub const MAX_LIMIT_VALUE: u32 = 10_000;
+
+#[derive(Clone, Encode, Decode, RuntimeDebug, MaxEncodedLen, TypeInfo, Eq, PartialEq)]
+#[scale_info(skip_type_params(T))]
+pub struct TradeVolumeLimit<T: Config> {
+	pub amount_in: T::Balance,
+	pub amount_out: T::Balance,
+	pub limit: T::Balance,
+}
+
+impl<T: Config> TradeVolumeLimit<T>
+where
+	T::Balance: PartialOrd,
+{
+	pub fn update_amounts(&mut self, amount_in: T::Balance, amount_out: T::Balance) -> DispatchResult {
+		self.amount_in = self
+			.amount_in
+			.checked_add(&amount_in)
+			.ok_or(ArithmeticError::Overflow)?;
+		self.amount_out = self
+			.amount_out
+			.checked_add(&amount_out)
+			.ok_or(ArithmeticError::Overflow)?;
+		Ok(())
+	}
+
+	pub fn check_min_limit(&self) -> DispatchResult {
+		if self.amount_out > self.amount_in {
```

Comment:
This one (`check_min_limit`) checks the lower limit (if we removed more liquidity than we can). And it can only happen if `amount_out > amount_in`. We actually test that `abs(amount_in - amount_out) <= limit`. And by dividing it into two checks we can tell whether we reached the lower or upper limit. There is another method (`check_max_limit`) to check the upper limit. We always call both methods, but the logic is divided into two cases.

Reply:
got it, thanks for the explanation!